### PR TITLE
Tdl 14963 removed endpoints dict

### PR DESCRIPTION
**Pull request by namrata270998**
# Description of change

- removed the endpoints dictionary
- added an Endpoints Class

# Manual QA steps
 - Check if the endpoint API URL is correct for each stream.
 - Check that the new code generates the same catalog files as before
 - Check that the new code generates the same records with and without the state file as before.

 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

**Review by dbshah1212 — APPROVED** [review]

**Comment by dbshah1212** [comment]
@namrata270998 Make sure you merge this code after back merging the tap-tester code and all tests passed.

File: tap_pendo/streams.py [hunk 1]
```diff
@@ -206,14 +75,31 @@ def update_currently_syncing(state, stream_name):
 class Server42xRateLimitError(Exception):
     pass
 
+
+class Endpoints:
```
Comment (karanpanchal-crest) [inline tap_pendo/streams.py:79]: @namrata270998 Add required unit test cases for these code changes

Comment (namrata270998) [reply tap_pendo/streams.py:79]: Done

File: tap_pendo/streams.py [hunk 2]
```diff
@@ -206,14 +75,31 @@ def update_currently_syncing(state, stream_name):
...
+    params = {}
+
+    def __init__(self, endpoint, method, headers=None, params=None):
+        self.endpoint = endpoint
+        self.method = method
+        self.headers = headers
+        self.params = params
+
+    def get_url(self, **kwargs):
+        return BASE_URL + self.endpoint.format(**kwargs)
```
Comment (karanpanchal-crest) [inline tap_pendo/streams.py:92]: @namrata270998 Add comments to this function

File: tap_pendo/streams.py [hunk 3]
```diff
@@ -206,14 +75,31 @@ def update_currently_syncing(state, stream_name):
...
+
 class Stream():
     name = None
     replication_method = None
     replication_key = None
     key_properties = KEY_PROPERTIES
     stream = None
-    method = "GET"
     period = None
+    endpoint = Endpoints("/api/v1/aggregation", "POST")
```
Comment (karanpanchal-crest) [inline tap_pendo/streams.py:102]: @namrata270998 Add comment to this code

File: tap_pendo/streams.py [hunk 4]
```diff
@@ -531,7 +415,6 @@ class Accounts(Stream):
     replication_method = "INCREMENTAL"
     replication_key = "lastupdated"
     key_properties = ["account_id"]
-    method = "POST"
```
Comment (karanpanchal-crest) [inline tap_pendo/streams.py:534]: @namrata270998 Why did we remove this attribute? Isn't it used anywhere?

File: tap_pendo/streams.py [hunk 5]
```diff
@@ -959,7 +852,6 @@ class Visitors(LazyAggregationStream):
     replication_method = "INCREMENTAL"
     replication_key = "lastupdated"
     key_properties = ["visitor_id"]
-    method = "POST"
 
     def get_endpoint(self):
```
Comment (karanpanchal-crest) [inline tap_pendo/streams.py:856]: @namrata270998 Is this function being used anywhere?

File: tap_pendo/streams.py [hunk 6]
```diff
@@ -1024,10 +917,12 @@ def sync(self, state, start_date=None, key_id=None):
     def get_fields(self):
         return self.request(self.name, json=self.get_body())
 
+
 class MetadataVisitors(Stream):
     name = "metadata_visitors"
     replication_method = "FULL_TABLE"
     key_properties = []
+    endpoint = Endpoints("/api/v1/metadata/schema/visitor", "GET")
```
Comment (karanpanchal-crest) [inline tap_pendo/streams.py:925]: @namrata270998 Please add unit test case for this code change

File: tap_pendo/streams.py [hunk 7]
```diff
@@ -1003,6 +895,7 @@ class MetadataAccounts(Stream):
     name = "metadata_accounts"
     replication_method = "FULL_TABLE"
     key_properties = []
+    endpoint = Endpoints("/api/v1/metadata/schema/account", "GET")
```
Comment (karanpanchal-crest) [inline tap_pendo/streams.py:898]: Add unit test case for this code change

File: tap_pendo/streams.py [hunk 8]
```diff
@@ -921,6 +806,14 @@ class VisitorHistory(Stream):
...
     key_properties = ['visitor_id']
     DATE_WINDOW_SIZE = 1
+    headers = {
+        'content-type': 'application/x-www-form-urlencoded'
+    }
+    params = {
+        "starttime": "start_time"
+    }
+    endpoint = Endpoints(
+        "/api/v1/visitor/{visitorId}/history", "GET", headers, params)
```
Comment (karanpanchal-crest) [inline tap_pendo/streams.py:816]: Add comments to this code change.
Also, can you please explain, how would the 'VisitorId' parameter be assigned a value?

File: tap_pendo/streams.py [hunk 9]
```diff
@@ -898,6 +781,7 @@ class Reports(Stream):
     name = "reports"
     replication_method = "INCREMENTAL"
     replication_key = "lastUpdatedAt"
+    endpoint = Endpoints("/api/v1/report", "GET")
```
Comment (karanpanchal-crest) [inline tap_pendo/streams.py:784]: Add unit test case for this code change

File: tests/unittests/test_endpoints_dict.py [hunk 10]
```diff
@@ -0,0 +1,18 @@
+from tap_pendo.streams import Endpoints, Stream
```
Comment (karanpanchal-crest) [inline tests/unittests/test_endpoints_dict.py:1]: @namrata270998 Did we execute the integration test cases for the same?

**Review by karanpanchal-crest — APPROVED** [review]

File: tap_pendo/streams.py (hunk 9 above)
Comment (namrata270998) [reply tap_pendo/streams.py:784]: Have added one common unit test case for this change in the test_endpoints_dict.py file

File: tap_pendo/streams.py (hunk 7 above)
Comment (namrata270998) [reply tap_pendo/streams.py:908]: No, removed it

File: tap_pendo/streams.py (hunk 5 above)
Comment (namrata270998) [reply tap_pendo/streams.py]: No, removed it

File: tap_pendo/streams.py (hunk 8 above)
Comment (namrata270998) [reply tap_pendo/streams.py:828]: Done.
The get_url function of the Endpoints class will be used, and in that function this `visitorId` will be formatted and replaced with the original one

**Pull request closed without merging**